depth_to_pointcloud/src/depth_to_pointcloud_node.cpp

```diff
@@ -0,0 +1,460 @@
+  double unit_scaling = DepthTraits<T>::toMeters( T(1) );
+  float constant_x = unit_scaling / model.fx();
+  float constant_y = unit_scaling / model.fy();
+  float bad_point = std::numeric_limits<float>::quiet_NaN();
+
+  PointCloud2Iterator<float> iter_x(*cloud_msg, "x");
+  PointCloud2Iterator<float> iter_y(*cloud_msg, "y");
+  PointCloud2Iterator<float> iter_z(*cloud_msg, "z");
+  const T* depth_row = reinterpret_cast<const T*>(&depth_msg->data[0]);
+  int row_step = depth_msg->step / sizeof(T);
+  for (int v = 0; v < (int)cloud_msg->height; ++v, depth_row += row_step)
+  {
+    for (int u = 0; u < (int)cloud_msg->width; ++u, ++iter_x, ++iter_y, ++iter_z)
+    {
+      T depth = depth_row[u];
+
+      // Missing points denoted by NaNs
+      if (!DepthTraits<T>::valid(depth))
+      {
+        if (range_max != 0.0)
+        {
+          depth = DepthTraits<T>::fromMeters(range_max);
+        }
+        else
+        {
+          *iter_x = *iter_y = *iter_z = bad_point;
+          continue;
+        }
+      }
+
+      // Fill in XYZ
+      *iter_x = (u - center_x) * depth * constant_x;
+      *iter_y = (v - center_y) * depth * constant_y;
+      *iter_z = DepthTraits<T>::toMeters(depth);
+    }
+  }
+}
+
+void depthCb(const sensor_msgs::msg::Image::SharedPtr image)
+{
+  if (cam_info == nullptr)
+  {
+    // we haven't gotten the camera info yet, so just drop until
+    // we do.
+    fprintf(stderr, "No camera info, skipping point cloud conversion\n");
+    return;
+  }
+  sensor_msgs::msg::PointCloud2::SharedPtr cloud_msg = std::make_shared<sensor_msgs::msg::PointCloud2>();
+  cloud_msg->header = image->header;
+  cloud_msg->height = image->height;
+  cloud_msg->width = image->width;
+  cloud_msg->is_dense = false;
+  cloud_msg->is_bigendian = false;
+  cloud_msg->fields.clear();
+  cloud_msg->fields.reserve(1);
+
+  sensor_msgs::msg::PointField pfx;
+  pfx.name = "x";
+  pfx.count = 1;
+  pfx.datatype = sensor_msgs::msg::PointField::FLOAT32;
+  pfx.offset = 0;
+  cloud_msg->fields.push_back(pfx);
+
+  sensor_msgs::msg::PointField pfy;
+  pfy.name = "y";
+  pfy.count = 1;
+  pfy.datatype = sensor_msgs::msg::PointField::FLOAT32;
+  pfy.offset = 4;
+  cloud_msg->fields.push_back(pfy);
+
+  sensor_msgs::msg::PointField pfz;
+  pfz.name = "z";
+  pfz.count = 1;
+  pfz.datatype = sensor_msgs::msg::PointField::FLOAT32;
+  pfz.offset = 8;
+  cloud_msg->fields.push_back(pfz);
+
+  cloud_msg->point_step = 16;
+  cloud_msg->row_step = cloud_msg->width * cloud_msg->point_step;
+  cloud_msg->data.resize(cloud_msg->height * cloud_msg->row_step);
+
+  // info_msg here is a sensor_msg::msg::CameraInfo::ConstSharedPtr; we should be able to get this from the astra_driver layer.
+  image_geometry::PinholeCameraModel model_;
+  model_.fromCameraInfo(cam_info);
+
+  convert<float>(image, cloud_msg, model_);
+
+  cloud_msg->header.frame_id = std::string("openni_depth_optical_frame");
```

Comment:
This hasn't been addressed. Please use a technology agnostic name because this node is not tied to openni AFAIK.
e.g. [image_depth_proc](http://wiki.ros.org/depth_image_proc) uses `/depth_optical_frame` and `/rgb_optical_frame`